Readme Revisions (#3)
SETO2243/forecasting: README.md

# Time Series Models

A library for composable forecasting models built on top of [scikit-learn](https://scikit-learn.org/stable/)

> This software library is the open source package for the work done in task 2: Endpoint data process and Analysis. The

Reviewer: ```suggestion
This software library is the open source package for the work done for [SETO project 2243](https://www.energy.gov/eere/solar/seto-2020-artificial-intelligence-applications-solar-energy#:~:text=for%20unobservable%20areas.-,Camus%20Energy,-Project%20Name%3A), _Improving Grid Awareness by Empowering Utilities with Machine Learning and Artificial Intelligence_, specifically task 2: _Endpoint data process and Analysis_. The
```

P3

package includes meter level forecasting and gap filling for customer load as well as photovoltaic (PV) generation 
forecast models for short term and intra-day PV forecasting.

The package provides the tools to construct machine learning models that fill gaps or forecast in the verification
datasets. This includes loading the data and applying the XGBoost estimator as described in section 3.1 of the Final Technical
Review (FTR). The composable model framework described below provides the configurable model inputs required for the 
Neighbor Informed Estimates and Community Analytics. The PV System model in section 3.2 of the FTR is also implemented using
the composable model framework. Example PV model construction is shown below.

> <!-- Add intro that provides project context. -->
> <!-- FTR - final technical report. Everything must link to there - except for this doc. Link to task numbers -->
> <!-- TODO: Get permalink for FTR once published and link it from here! -->
> 
> <!-- Explain what this does for lawyers... it solves this particular problem for the FTR - as well as this class of problem in general -->
> <!-- Expand on connections to SOPO: be clear about mapping to the things we completed. Make sure keyword searches will work -->
> <!-- Read SOPO: look for nouns in the relevant task descriptions. -->

Reviewer: ```suggestion
<!-- TODO: Get permalink for FTR once published and link it from here! -->
```

Looks like this are mostly done 👏

P3

The API for fit, predict, and metrics is reduced to specifying a start and end times for a given location.
The model must construct feature data using column transforms. Having done so, forecasting as a service become trivial.

## Installation

> <!-- Talk about env for this to run - should be a complete description - provide the context. -->
> 

Reviewer: ```suggestion
```

✅

P3

block. The estimator can be either a Classifier (a discrete estimator) or a Regressor (a continuous estimator). There 
are many different numerical techniques for [supervised learning](https://scikit-learn.org/stable/supervised_learning.html)
estimators. The process is the last essential component. It defines the timeseries being forecast and the available 
feature data that might have predictive value. Having composed a Model class from these three parts, it is then up to 
the user to create an instance of the class with configuration arguments that tune the model features for the specific
meter load or pv forecast. 

<!-- Explain what this does - for lawyers... point to ipynb example -->
<!-- Consider adding sklearn model diagram if it would help (rather than hurt) the explanation -->

### Compose a model class

New models are defined as Python classes, which utilize building blocks provided by this library as base classes. For example, here is the beginning of a model using an Ordinary Least Squares estimator to forecast Balancing Area Hourly Load:

```python
class OlsModel(BalancingAreaHourly, LinearRegressor, RegularTimeSeriesModel):
  ...
```

> And this example is model using an [XgBoost](https://xgboost.readthedocs.io/en/stable/) estimator to forecast AMI (meter) Hourly Load:

Reviewer: ```suggestion
And this example is model using an [XgBoost](https://xgboost.readthedocs.io/en/stable/) estimator to forecast AMI (smart meter) Hourly Load:
```

P4

<!-- Expand on this - show examples -->

### PV Model
The PV Model as described in the FTR section 3.2 uses the same composable framework define models
using the HRRR weather (see below) as an input to the NREL PySam PV generatin algorithm. For the 
project we used the PySam generation forecast directly using the configration shown below with the
IdentityRegressor. Building additional input features for sites with direct telemetry would allow
using machine learning models like xgboost too.

```python
pv_config = dict(
    site_config_mapping="RESOURCE_SELF",
    site_meter_mapping=None,
    site_latlong_mapping="RESOURCE_SELF",
    source_mode="12_hour_horizon",
    lags=None,
)

class PVForecastModel(
    PVForecast,
    IdentityRegressor,
    RegularTimeSeriesModel,
):
    pass

pv_instance = PVForecastModel(**pv_config)
pv_instance.model
```

### Usage
Engineers and data scientists commonly use an interactive web-based development environment called [Jupyter Notebook](https://jupyter.org/)
(now Jupyter Lab) to explore and visualize data and algorithms in a cell based execution environment. 

> <!-- Developers commonly use ipython notebooks for data science and... -->

Reviewer: ```suggestion
```

✅

P3

> A library for composable forecasting models built on top of scikit-learn

Reviewer: ```suggestion
A library for composable forecasting models built on top of [scikit-learn](https://scikit-learn.org/stable/)
```

Yay for links.

P3

<!-- Add intro that provides project context. -->
<!-- FTR - final technical report. Everything must link to there - except for this doc. Link to task numbers -->
<!-- TODO: Get permalink for FTR once published and link it from here! -->

<!-- Explain what this does for lawyers... it solves this particular problem for the FTR - as well as this class of problem in general -->
<!-- Expand on connections to SOPO: be clear about mapping to the things we completed. Make sure keyword searches will work -->
<!-- Read SOPO: look for nouns in the relevant task descriptions. -->

The API for fit, predict, and metrics is reduced to specifying a start and end times for a given location.
The model must construct feature data using column transforms. Having done so, forecasting as a service become trivial.

## Installation

<!-- Talk about env for this to run - should be a complete description - provide the context. -->

> This library is designed for use by technical engineers and data scientists. It takes advantage of the python
> data science ecosystem and therefore requires installation of many third party open source libraries. It has 
> been developed and tested in a linux operating system. Running on a docker container such as 
> the [canonical Ubuntu image](https://hub.docker.com/_/ubuntu) is strongly recommended. The library was 
> developed using Ubuntu 22.04 (Jammy) with Python 3.10.6.

Reviewer: ```suggestion
This library is designed for use by technical engineers and data scientists. It takes advantage of the Python
data science ecosystem and therefore requires installation of many third party open source libraries. It has 
been developed and tested in a Linux operating system. Running on a Docker container such as 
the [canonical Ubuntu image](https://hub.docker.com/_/ubuntu) is strongly recommended. The library was 
developed using Ubuntu 22.04 (Jammy) with Python 3.10.6.
```

Proper nouns.

P3

> ### Installing system libraries

Reviewer: Maybe include a Docker command for pulling/running your suggested container, so readers can just cut-paste. For instance:

````suggestion
### Installing system libraries

After installing [Docker](https://docs.docker.com/engine/install/), run the following command to setup a basic Jammy container with this library:

```sh
docker run -it ubuntu:jammy /bin/bash
apt-get update
apt-get install git
git clone https://github.com/SETO2243/forecasting.git
cd /forecasting
```
````

P3

> Reading the grib weather data requires the eccodes library: `apt-get install libeccodes-dev`

Reviewer: What is grib weather data and why is it important? (Concept not previously introduced.)

````suggestion
Reading the grib weather data requires the eccodes library, which is available from the Jammy package repo, installable via shell:

```sh
apt-get install libeccodes-dev
```
````

Partially making it clear that this is a shell command.

P3

> ### Installing python libraries
> `pip3 install -r requirements.txt`
> 
> Verify you can run the unit tests `python -m unittest`

Reviewer: ````suggestion
### Installing Python libraries

Run the following command to install Python package dependencies:

```sh
pip3 install -r requirements.txt
```

### Validating your environment

Verify that your environment is fully functional by running the automated unit tests:

```sh
python -m unittest
```

This will print "SUCCESS" near the end if the code work correctly in your new environment.
````

Moar clarity!

P3

> To start jupyter notebook run `jupyter notebook --NotebookApp.ip=0.0.0.0`
> Now you can open the example notebook and execute the cells in the demonstration.

Reviewer: ````suggestion
To start jupyter notebook run:

```sh
jupyter notebook --NotebookApp.ip=0.0.0.0
```

This will print a URL, which you can open in your browser. Then open the example notebook and execute the cells in the demonstration to get acquainted with the functionality.
````

P3

<!-- Phase 2 (after FTR accepted): add docker container in July -->

## Usage
Models can be composed of mixins for various estimators and forecast processes. These composable

> pieces can be put together in different ways to solve many problems. The RegularTimeSeriesModel is the base.

Reviewer: Diagram of the simplest possible complete composition might be helpful here, as illustration.

P3

> pieces can be put together in different ways to solve many problems. The RegularTimeSeriesModel is the base.
> The core that problem specific parts are added to when forecasting or gap filling a particular timeseries.

Reviewer: ```suggestion
pieces can be put together in different ways to solve many problems. The RegularTimeSeriesModel is the
core that problem specific parts are added to when forecasting or gap filling a particular timeseries.
```

Otherwise second thing doesn't appear to be a sentence.

P3

> The estimator is the next essential building block. It can be a Classifier or a Regressor. There are many 

Reviewer: Should define classifiers and regressors, or link to a definition.

P3

Author: Aded more links to sklearn

different numerical techniques that can be applied via the estimator. The process is the last essential 
component. It defines the timeseries being forecast and the available feature data that might have predictive value.

> Having composed a Model class from these three parts, it is then up to the user to create an instance of the class that 

Reviewer: ```suggestion
Having composed a Model class from these three parts, it is then up to the user to create an instance of the class 
```

P3

with configuration arguments that tune the model features for the specific meter load or pv forecast. 

<!-- Explain what this does - for lawyers... point to ipynb example -->
<!-- Consider adding sklearn model diagram if it would help (rather than hurt) the explanation -->

### Compose a model class

> A model using an Ordinary Least Squares estimator to forecast a Balancing Area Hourly Load
> ```python
> class NpOlsModel(BalancingAreaHourly, LinearRegressor, RegularTimeSeriesModel):
>   pass
> ```

Reviewer: ````suggestion
New models are defined as Python classes, which utilize building blocks provided by this library as base classes. For example, here is the beginning of a model using an Ordinary Least Squares estimator to forecast Balancing Area Hourly Load:

```python
class NpOlsModel(BalancingAreaHourly, LinearRegressor, RegularTimeSeriesModel):
  ...
```
````

Also maybe explain what "NP" is.

Also also, is an `import` statement necessary here, that we should include for completeness?

P3

> A model using an XgBoost estimator to forecast AMI (meter) Hourly Load

Reviewer: ```suggestion
And this example is model using an [XgBoost](https://xgboost.readthedocs.io/en/stable/) estimator to forecast AMI (meter) Hourly Load:

```

P3

```python
class XgbModel(AmiHourlyForecast, XgbRegressor, RegularTimeSeriesModel):
  pass
```

> Additional behaviors like filters can also be added via composition. See the notebook demo for examples.

Reviewer: What are the inputs and outputs of filters? What's an example use case for filters?

P2

Author: Removing nan/missing values in the training set.
Linear interp/extrapolation of missing/nan feature data.

These are included in the repo. I added a link to the module.

### Configure a model instance

> The initialization arguments control feature construction. The configuration arguments are specific to the

Reviewer: What is a feature? And what is its construction?

P3

Author: Added more context. Want to avoid going to deep here. This could get very long.

process and the estimator used to compose the model.

As an example, we can configure XgbModel from above with three types of features: lagged correlations, one hot encoded 

> day-of-week values and a harmonic feature that decomposes time into a sine and cosine waves with the specified frequencies.

Reviewer: Hmmm, I like that harmonic feature is defined here. Maybe also define lagged correlation and one hot encoded day-of-week. Maybe put these in a bulleted list.

P3

Author: Added links and reference to the FTR for now - let me know if you think more in the readme would help.

```python
config = dict(
lags = np.array([24, 48, 168], dtype="timedelta64[h]"),
day_of_week=True,
harmonics = np.array([24, 168, 365*24], dtype="timedelta64[h]")
)
instance = XgbModel(**config)
```

### Fit & Predict
Once a model instance is created, we must train the machine learning algorithm. The fit method takes
a start date, stop date and a list of one or more identifiers. For process classes that support it, this

> allows training a single model for a cohort of resources.

Reviewer: Do some process classes simply not allow any learning? If not, why would they not support training?

P3

Author: Ah - gotcha. Will fix the sentence structure. All classes implement fit, but only some support cohorts.

```python
instance.fit("2021-01-15", "2021-06-01", "55797646")
```

Once the model instance is trained, we can call the `predict` method (again with a start, stop and one or more
resource identifiers) to generate new predicted values from the input features for the specified date range.
```python
instance.predict_dataframe(
    "2021-03-15",
    "2021-03-15T04",
    "55797646",
    range=True,
)
```
![Predicted & Actual Values](https://github.com/SETO2243/forecasting/assets/84335963/99cdaaf1-d15a-40fd-bbfa-534da6de4b49)

<!-- Expand on this - show examples -->

### Usage

> Engineers and data scientists commonly use an interactive web-based development environment called [Jupyter Notebook](https://jupyter.org/)
> (now jupyter lab) to explore and visualize data and algorithms in a cell based execution environment. 
> <!-- Developers commonly use ipython notebooks for data science and... -->
> 
> An [example notebook](https://github.com/SETO2243/forecasting/blob/main/example.ipynb) is provided in this github
> repository which demonstrates the core capabilities of the time series models library developed for the SETO project. 

Reviewer: ```suggestion
Engineers and data scientists commonly use an interactive web-based development environment called [Jupyter Notebook](https://jupyter.org/)
(now Jupyter Lab) to explore and visualize data and algorithms in a cell based execution environment. 
<!-- Developers commonly use ipython notebooks for data science and... -->

An [example notebook](https://github.com/SETO2243/forecasting/blob/main/example.ipynb) is provided in this GitHub
repository which demonstrates the core capabilities of the time series models library developed for the SETO project. 
```

Proper nouns.

P3

> 

Reviewer: Maybe include a screen shot here.

P3

### High Resolution Rapid Refresh weather data
The [High Resolution Rapid Refresh](https://rapidrefresh.noaa.gov/hrrr/) (HRRR) forecast is an operational weather 
forecasting product of the NOAA Center for Environmental Prediction. The HRRR forecast results are publicly available 
on multiple cloud vendor platforms ([AWS](https://registry.opendata.aws/collab/noaa/),
[GCP](https://cloud.google.com/blog/products/data-analytics/noaa-datasets-on-google-cloud-for-environmental-exploration),
[Azure](https://planetarycomputer.microsoft.com/catalog)) via a public private partnership with the
[NOAA Open Data Dissemination](https://www.noaa.gov/information-technology/open-data-dissemination) program.

These cloud providers host the petabyte scale archive of grib2 files created by the hourly HRRR operational forecast system.
As part of the project a set of metadata file that index the archive was created using an open source tool
called [Kerchunk](https://github.com/fsspec/kerchunk). The weather data is a public archive provided by NODD. To use it
the timeseries models developed for the project a kerchunk metadata index must be created. A sample index is provided

> in the project [GCP bucket](https://console.cloud.google.com/storage/browser/seto2243-forecasting).

Reviewer: ```suggestion
in the project [GCS bucket](https://console.cloud.google.com/storage/browser/seto2243-forecasting).
```

P4

## Contents
<!-- Add context: domain specific technical langauge to help engineers find specific functionality in the repository -->

> The following contents uses domain specific language to help engineers (users) find specific functionality implemented by the library 

Reviewer: ```suggestion
Library code is organized into a number of subpackages described below, to aid engineers working on writing or debugging code using this library:
```

P3
